provider/vsphere/internal/vsphereclient/client.go

```diff
@@ -456,23 +458,65 @@ func (c *Client) destroyVM(
 
 func (c *Client) cloneVM(
 	ctx context.Context,
+	args CreateVirtualMachineParams,
 	srcVM *object.VirtualMachine,
 	dstName string,
 	vmFolder *object.Folder,
+	datacenter *object.Datacenter,
 	taskWaiter *taskWaiter,
 ) (*object.VirtualMachine, error) {
+	configSpec := &types.VirtualMachineConfigSpec{}
+
+	if args.Constraints.HasCpuCores() {
+		configSpec.NumCPUs = int32(*args.Constraints.CpuCores)
+	}
+	if args.Constraints.HasMem() {
+		configSpec.MemoryMB = int64(*args.Constraints.Mem)
+	}
+	if args.Constraints.HasCpuPower() {
+		cpuPower := int64(*args.Constraints.CpuPower)
+		configSpec.CpuAllocation = &types.ResourceAllocationInfo{
+			Limit:       &cpuPower,
+			Reservation: &cpuPower,
+		}
+	}
+	if configSpec.Flags == nil {
+		configSpec.Flags = &types.VirtualMachineFlagInfo{}
+	}
+
 	task, err := srcVM.Clone(ctx, vmFolder, dstName, types.VirtualMachineCloneSpec{
-		Config:   &types.VirtualMachineConfigSpec{},
-		Location: types.VirtualMachineRelocateSpec{},
+		Config: configSpec,
+		Location: types.VirtualMachineRelocateSpec{
+			Pool: &args.ResourcePool,
+		},
 	})
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	info, err := taskWaiter.waitTask(ctx, task, "cloning VM")
 	if err != nil {
-		return nil, err
+		return nil, errors.Trace(err)
 	}
-	return object.NewVirtualMachine(c.client.Client, info.Result.(types.ManagedObjectReference)), nil
+
+	vm := object.NewVirtualMachine(c.client.Client, info.Result.(types.ManagedObjectReference))
+
+	if args.Constraints.RootDisk != nil {
+		// The user specified a root disk, so extend the VM's
+		// disk before powering the VM on.
+		args.UpdateProgress(fmt.Sprintf(
+			"extending disk to %s",
+			humanize.IBytes(*args.Constraints.RootDisk*1024*1024),
```

Comment:
I think there's a helper we can use that doesn't require 1024 * 1024

Reply:
Should be humanize.IBytes(*args.Constraints.RootDisk * humanize.GiByte)

Reply:
Thanks @hpidcock. Updated.